builtin/credential/aws/path_login.go

```diff
@@ -888,6 +1048,389 @@ func (b *backend) pathLoginRenew(
+	}
+
+	endpoint := "https://sts.amazonaws.com"
+
+	if config != nil {
+		if config.IAMServerIdHeaderValue != "" {
+			err = validateVaultHeaderValue(headers, parsedUrl, config.IAMServerIdHeaderValue)
+			if err != nil {
+				return logical.ErrorResponse(fmt.Sprintf("error validating %s header: %s", iamServerIdHeader, err)), nil
+			}
+		}
+		if config.STSEndpoint != "" {
+			endpoint = config.Endpoint
+		}
+	}
+
+	clientArn, accountID, err := submitCallerIdentityRequest(method, endpoint, parsedUrl, body, headers)
+	if err != nil {
+		return logical.ErrorResponse(fmt.Sprintf("error making upstream request: %s", err)), nil
+	}
+	canonicalArn, principalName, sessionName, err := parseIamArn(clientArn)
+	if err != nil {
+		return logical.ErrorResponse(fmt.Sprintf("Error parsing arn: %s", err)), nil
+	}
+
+	roleName := data.Get("role").(string)
+	if roleName == "" {
+		roleName = principalName
+	}
+
+	roleEntry, err := b.lockedAWSRole(req.Storage, roleName)
+	if err != nil {
+		return nil, err
+	}
+	if roleEntry == nil {
+		return logical.ErrorResponse(fmt.Sprintf("entry for role %s not found", roleName)), nil
+	}
+
+	if !strutil.StrListContains(roleEntry.AllowedAuthTypes, iamAuthType) {
+		return logical.ErrorResponse(fmt.Sprintf("auth method iam not allowed for role %s", roleName)), nil
+	}
+
+	// The role creation should ensure that either we're inferring this is an EC2 instance
+	// or that we're binding an ARN
+	if roleEntry.BoundIamPrincipalARN != "" && roleEntry.BoundIamPrincipalARN != canonicalArn {
+		return logical.ErrorResponse(fmt.Sprintf("IAM Principal %q does not belong to the role %q", clientArn, roleName)), nil
+	}
+
+	policies := roleEntry.Policies
+	rTagMaxTTL := time.Duration(0)
+
+	inferredEntityType := ""
+	inferredEntityId := ""
+	if roleEntry.RoleInferredType == ec2EntityType {
+		instance, err := b.validateInstance(req.Storage, sessionName, roleEntry.InferredAWSRegion, accountID)
+		if err != nil {
+			return logical.ErrorResponse(fmt.Sprintf("failed to verify %s as a valid EC2 instance in region %s", sessionName, roleEntry.InferredAWSRegion)), nil
+		}
+
+		// build a fake identity doc to pass on metadata about the instance to verifyInstanceMeetsRoleRequirements
+		identityDoc := &identityDocument{
+			Tags:        nil, // Don't really need the tags, so not doing the work of converting them from Instance.Tags to identityDocument.Tags
+			InstanceID:  *instance.InstanceId,
+			AmiID:       *instance.ImageId,
+			AccountID:   accountID,
+			Region:      roleEntry.InferredAWSRegion,
+			PendingTime: instance.LaunchTime.Format(time.RFC3339),
+		}
+
+		roleTagResp, validationError, err := b.verifyInstanceMeetsRoleRequirements(req.Storage, instance, roleEntry, roleName, identityDoc)
+		if err != nil {
+			return nil, err
+		}
+		if validationError != nil {
+			return logical.ErrorResponse(fmt.Sprintf("Error validating instance: %s", validationError)), nil
+		}
+
+		if roleTagResp != nil {
```

Comment:
I see what you are saying and it makes sense too. However, I still have some concerns.

Yes, role tags do the job of authorization, but it it tightly coupled with the ec2 auth mechanism. It has a `nonce`, `disable_reauthentication` and `allow_instance_migration` fields which are only applicable when authentication is done via identity document. Even if we come up with a way to create role tags based on auth type, by nature, role tags are immutable and are not even stored at the backend and hence there can't be any upgrade paths for existing role tags in terms of being able to distinguish them from belonging to ec2 and/or iam roles.

For those who never use ec2 auth mechanism in future, the fields in role tag might be hard to reason about. Role tags are designed to contain items that restrict the capabilities of the role. If the role has a bound by IAM principal ARN and when the authentication is not performed using identity document, the role tag fields does not represent the subset of capabilities on the role.

I think there are two ways to go forward:
1) When the role is bound by IAM principal ARN, make sure that role tag does not contain items that are relevant to the identity document.
2) When the role is bound by IAM principal ARN, don't allow role tags to be created at all.

I am inclined towards the 2nd option though. Further in this direction, to be able to reason about what's going on at any given time, I feel that the role should be tied to any *one* auth type. For example, when the auth types contain both iam and ec2, which credentials do we prioritize if both identity document and the new ones are supplied? It also gets harder to determine when we should allow role tags and when not.

Here are my thoughts on what we should do. Let me know what you think.
1) One role should allow only one type of authentication, ec2 or iam.
2) When the type is ec2, everything remains the same as it was before.
3) When the type is iam, accept new credentials. If inferred type is ec2_instance, apply the binds belonging to ec2 instances but not those that are specific to identity document. Specifically keep the role tag processing out.